### PR TITLE
Assume simple model interface in adversary

### PR DESCRIPTION
**Pull request by mzweilin**
# What does this PR do?

This PR makes `Adversary` work with any model that `output_dict = model(batch)`. `LitModular` and `SequentialDict` are not required any more.

- Add `attack_step()` in `LitModular`.
- `Adversary` works with models that are not `LitModular`.
- Add `batch_converter` in `Adversary` to standardize input for adversary. (Anomalib gives ImageNet normalized input.)

This PR depends on #185

This PR is split into #210 and #215

## Type of change

Please check all relevant options.

- [x] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] `pytest`
- [x] `CUDA_VISIBLE_DEVICES=0 python -m mart experiment=CIFAR10_CNN_Adv trainer=gpu trainer.precision=16` reports 70% (21 sec/epoch).
- [x] `CUDA_VISIBLE_DEVICES=0,1 python -m mart experiment=CIFAR10_CNN_Adv trainer=ddp trainer.precision=16 trainer.devices=2 model.optimizer.lr=0.2 trainer.max_steps=2925 datamodule.ims_per_batch=256 datamodule.world_size=2` reports 70% (14 sec/epoch).

## Before submitting

- [x] The title is **self-explanatory** and the description **concisely** explains the PR
- [x] My **PR does only one thing**, instead of bundling different changes together
- [ ] I list all the **breaking changes** introduced by this pull request
- [ ] I have commented my code
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have run pre-commit hooks with `pre-commit run -a` command without errors

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: mart/attack/adversary.py [hunk 1]
```diff
@@ -131,57 +150,43 @@ def training_step(self, batch, batch_idx):
...
-        # do not know where the Adversary lives inside the model, we also need the
-        # remaining sequence to be able to get a loss.
-        if model:
-            self._attack(**batch)
-
-        perturbation = self.perturber(**batch)
-        input_adv = self.composer(perturbation, **batch)
+    def forward(self, *, input, target, model):
+        # What we need is a frozen model that returns (a dictionary of) logits, or losses.
+        self.model = model
```
Comment (dxoigmn) [inline mart/attack/adversary.py:168]: This is a bad idea...

Comment (mzweilin) [reply mart/attack/adversary.py]: Could you elaborate? I made an commit to hide the model in [2cd7ca8](https://github.com/IntelLabs/MART/pull/200/commits/2cd7ca86bf1f8886b6f0e110347696d5fd908736)

Comment (mzweilin) [reply mart/attack/adversary.py]: Now I hide `model` in adversary's data batch.

Comment (dxoigmn) [reply mart/attack/adversary.py]: It's easier to reason about modules when they're purely functional. Modifying self means one needs to mentally keep track of state of the module now.

**Review by dxoigmn — CHANGES_REQUESTED** [review]
Can you make this PR more minimal? Lots going on here and I don't have the patience to review large PRs.

**Comment by dxoigmn** [comment]
Also, can you add why you made this change? Seems like a lot of changes for some unknown use case.

**Comment by mzweilin** [comment]
> Can you make this PR more minimal? Lots going on here and I don't have the patience to review large PRs.

Okay. I have merged main into #185 to hide many changes that are already in main.

File: mart/nn/nn.py [hunk 2]
```diff
@@ -57,7 +57,8 @@ def __init__(self, modules, sequences=None):
         self._sequences = {
             name: self.parse_sequence(sequence) for name, sequence in sequences.items()
         }
-        self._sequences[None] = self
+        # We intend to make training sequence as the default sequence.
+        # self._sequences[None] = self
```
Comment (dxoigmn) [inline mart/nn/nn.py:61]: This is a big semantic change! Can this not be changed?

File: mart/configs/experiment/CIFAR10_CNN_Adv.yaml [hunk 3]
```diff
@@ -1,7 +1,7 @@
 # @package _global_
 
 defaults:
-  - /attack@callbacks.adversarial_training.adversary: classification_eps1.75_fgsm
+  - /attack@callbacks.adversarial_training.train_adversary: classification_eps1.75_fgsm
```
Comment (dxoigmn) [inline mart/configs/experiment/CIFAR10_CNN_Adv.yaml:4]: Why the change?

File: mart/attack/adversary.py [hunk 4]
```diff
@@ -101,6 +101,9 @@ def __init__(
             assert self._attacker.max_epochs == 0
             assert self._attacker.limit_train_batches > 0
 
+        # TODO: Make this configurable. E.g. [0,1] <-> [0,255]
+        self.transform = self.untransform = lambda x: x
+
```
Comment (dxoigmn) [inline mart/attack/adversary.py:106]: Why is this necessary? How is this making things simpler?

File: mart/callbacks/adversarial_training.py [hunk 5]
```diff
@@ -37,32 +36,27 @@ def teardown(self, trainer, pl_module, stage=None):
...
-        # Create attacked model where the adversary executes before the model
-        # FIXME: Should we just use pl_module.training_step? Ideally we would not decompose batch
-        #        and instead pass batch directly to the underlying pl_module since it knows how to
-        #        interpret batch.
-        def attacked_model(input, **batch):
-            input_adv = adversary(input=input, **batch)
-            return pl_module(input=input_adv, **batch)
+        # Skip if adversary is not defined for the phase train/validation/test.
+        if adversary is None:
+            return batch
```
Comment (dxoigmn) [inline mart/callbacks/adversarial_training.py:51]: Why did you change this when this was already protected against above?

File: mart/callbacks/adversarial_training.py [hunk 6]
```diff
@@ -37,32 +36,27 @@ def teardown(self, trainer, pl_module, stage=None):
...
 
         # Move adversary to same device as pl_module and run attack
+        adversary.to(pl_module.device)
+
```
Comment (dxoigmn) [inline mart/callbacks/adversarial_training.py:55]: Why change this code?

File: mart/callbacks/adversarial_training.py [hunk 7]
```diff
@@ -37,32 +36,27 @@ def teardown(self, trainer, pl_module, stage=None):
...
-            return batch
 
-        # Create attacked model where the adversary executes before the model
-        # FIXME: Should we just use pl_module.training_step? Ideally we would not decompose batch
-        #        and instead pass batch directly to the underlying pl_module since it knows how to
-        #        interpret batch.
-        def attacked_model(input, **batch):
-            input_adv = adversary(input=input, **batch)
-            return pl_module(input=input_adv, **batch)
+        # Skip if adversary is not defined for the phase train/validation/test.
```
Comment (dxoigmn) [inline mart/callbacks/adversarial_training.py:49]: How does the adversary execute now?

**Review by dxoigmn — CHANGES_REQUESTED** [review]
There's still too much going on in this PR for me to review. There are several changes:
1) Logging of gain stuff
2) Transform/untransform
3) Getting rid of the use of step

Can you decouple those? I cannot for the life of me follow the logic in this PR.

File: mart/configs/experiment/CIFAR10_CNN_Adv.yaml (hunk 3 above)
Comment (mzweilin) [reply mart/configs/experiment/CIFAR10_CNN_Adv.yaml:4]: We don't run adversary in validation in this experiment.

File: mart/configs/callbacks/adversarial_training.yaml [hunk 8]
```diff
@@ -1,3 +1,3 @@
 adversarial_training:
   _target_: mart.callbacks.AdversarialTraining
-  adversary: ???
+  # adversary: ???
```
[the remark this answers is not in the document]
Comment (mzweilin) [reply mart/configs/callbacks/adversarial_training.yaml:3]: Sometimes we don't want adversary in validation.

**Comment by mzweilin** [comment]
This PR is split into https://github.com/IntelLabs/MART/pull/210 and https://github.com/IntelLabs/MART/pull/215
